network/newstream/sync_provider.go

```diff
@@ -0,0 +1,359 @@
+func syncSubscriptionsDiff(peerPO, prevDepth, newDepth, max int, syncBinsOnlyWithinDepth bool) (subBins, quitBins []int) {
+	newStart, newEnd := syncBins(peerPO, newDepth, max, syncBinsOnlyWithinDepth)
+	if prevDepth < 0 {
+		if newStart == -1 && newEnd == -1 {
+			return nil, nil
+		}
+		// no previous depth, return the complete range
+		// for subscriptions requests and nothing for quitting
+		return intRange(newStart, newEnd), nil
+	}
+
+	prevStart, prevEnd := syncBins(peerPO, prevDepth, max, syncBinsOnlyWithinDepth)
+	if newStart == -1 && newEnd == -1 {
+		// this means that we should not have any streams on any bins with this peer
+		// get rid of what was established on the previous depth
+		quitBins = append(quitBins, intRange(prevStart, prevEnd)...)
+		return
+	}
+
+	if newStart < prevStart {
+		subBins = append(subBins, intRange(newStart, prevStart)...)
+	}
+
+	if prevStart < newStart {
+		quitBins = append(quitBins, intRange(prevStart, newStart)...)
+	}
+
+	if newEnd < prevEnd {
+		quitBins = append(quitBins, intRange(newEnd, prevEnd)...)
+	}
+
+	if prevEnd < newEnd {
+		subBins = append(subBins, intRange(prevEnd, newEnd)...)
+	}
+
+	return subBins, quitBins
+}
+
+// syncBins returns the range to which proximity order bins syncing
+// subscriptions need to be requested, based on peer proximity and
+// kademlia neighbourhood depth. Returned range is [start,end), inclusive for
+// start and exclusive for end.
+// syncBinsOnlyWithinDepth toggles between having requested streams only within depth(true)
+// or rather with the old stream establishing logic (false)
+func syncBins(peerPO, depth, max int, syncBinsOnlyWithinDepth bool) (start, end int) {
+	if syncBinsOnlyWithinDepth && peerPO < depth {
+		// we don't want to request anything from peers outside depth
+		return -1, -1
+	}
+	if peerPO < depth {
+		// subscribe only to peerPO bin if it is not
+		// in the nearest neighbourhood
+		return peerPO, peerPO + 1
+	}
+	// subscribe from depth to max bin if the peer
+	// is in the nearest neighbourhood
+	return depth, max + 1
+}
+
+// intRange returns the slice of integers [start,end). The start
+// is inclusive and the end is not.
+func intRange(start, end int) (r []int) {
+	for i := start; i < end; i++ {
+		r = append(r, i)
+	}
+	return r
+}
+
+func checkKeyInSlice(k int, slice []int) (found bool) {
+	for _, v := range slice {
+		if v == k {
+			found = true
+		}
+	}
+	return
+}
+
+func (s *syncProvider) ParseKey(streamKey string) (interface{}, error) {
+	b, err := strconv.Atoi(streamKey)
+	if err != nil {
+		return 0, err
+	}
+	if b < 0 || b > 16 {
+		return 0, errors.New("stream key out of range")
+	}
+	return uint8(b), nil
+}
+
+func (s *syncProvider) EncodeKey(i interface{}) (string, error) {
+	v, ok := i.(uint8)
+	if !ok {
+		return "", errors.New("error encoding key")
+	}
+	return fmt.Sprintf("%d", v), nil
+}
+
+func (s *syncProvider) StreamName() string { return s.name }
+
+func (s *syncProvider) Boundedness() bool { return false }
```

Comment:
I fail to see why we need `Boundedness` and `Bound`. Why do we add such indirection without using it. If this is so important how come we don't have an implementation of it.

From the spec:
`when a client requests a bounded range, server should respond to the client range requests with either offered hashes (if roundtrip is required) or chunks (if not) or an end-of-batch message if there are no more to offer. If none of these responses arrive within a timeout interval, client must drop the upstream peer.`

We don't have an example of `Boundedness` or of `Roundtrip`, so I fail to see why this is useful, apart from introducing indirection in already not simple functionality. We've been working on `pull-sync` new version for months, the previous version has been buggy in production for probably more than a year (and still throwing subscription errors), but we still continue to add code prematurely so that we make our lives harder...

Reply:
@zelig @janos this comment is still applicable. @janos no action required from your side, but if you understand this, maybe you can explain it to me :)